### PR TITLE
feat: Skia for V4! 🥳 

### PR DESCRIPTION
**Pull request by mrousavy**
## What

Skia integration for V4! 🥳 

TL;DR: Allows you to "draw stuff onto the Camera Frame" in realtime.
This is different to just adding a view ontop of the `<Camera>`, because this actually draws inside the Frame buffer, meaning this can be extended to also be applied to captured photos and videos in the future. For now it's preview only.

For example, you can draw color filters, bounding boxes or circles for faces/qr-codes, face or license plate blurring, color corrections, animations on the frame, etc etc

### Backstory

After almost a year since the first idea and prototype of integrating Skia into VisionCamera, we can finally see it happen.
The main issues back then (July/August 2023) were that the react-native-skia dependency in VisionCamera's codebase was too harsh. It had a strong native dependency, we needed to add a bunch of native views including Metal/OpenGL Surface Views for preview, Choreographers for syncing, setting up Metal and OpenGL contexts, adding a native dependency and header include paths on libskia and all of the other skia companion libs, and we had a bunch of complicated nested branches in the Camera codebase where we switched logic between a normal Frame Processor and a Skia Frame Processor.
All of that is a thing from the past now, since I implemented a way to only expose a handle to the Frame ("platform buffer") to JS, and thanks to @wcandillon's amazing help (thank you so much for your time!) we got a `MakeImageFromPlatformBuffer(..)` function into react-native-skia.
Now we can create an `SkImage` from a `Frame`, straight from JS.

### API

In addition to `useFrameProcessor` ("readonly" Frame Processors) a new API has been introduced: `useSkiaFrameProcessor`.

In contrast to the `useFrameProcessor` hook, with the `useSkiaFrameProcessor` hook you now get `DrawableFrame`s instead of normal `Frame`s.

The `DrawableFrame` has a `render(..)` function to render itself (otherwise nothing gets rendered), and all other functions from `Frame` and `SkCanvas`.

For example, to render the image with a blur filter:

```tsx
const frameProcessor = useSkiaFrameProcessor((frame) => {
  'worklet'

  // 1. create blur filter
  const blurRadius = 10
  const blurFilter = Skia.ImageFilter.MakeBlur(blurRadius, blurRadius, TileMode.Repeat, null)

  // 2. wrap blur filter in Paint
  const paint = Skia.Paint()
  paint.setImageFilter(blurFilter)

  // 3. render image with the blur filter paint
  frame.render(paint)
}, [])
```

Or to draw a red rectangle on the middle of the Frame:

```tsx
const frameProcessor = useSkiaFrameProcessor((frame) => {
  'worklet'

  // 1. render frame normally
  frame.render() 

  // 2. create a red paint
  const redPaint = Skia.Paint()
  const red = Skia.Color('red')
  redPaint.setColor(red)

  // 3. draw a rectangle in the center of the screen
  const rect = Skia.XYWHRect(frame.width / 2, frame.height / 2, 150, 150)
  frame.drawRect(rect, redPaint)
}, [])
```

On Android, make sure to set the Camera's `pixelFormat={...}` to `"rgb"`. On iOS, you can use `"yuv"` or `"rgb"`, where `"yuv"` is typically more efficient.

> Note: Skia Frame Processors require react-native-reanimated (>= 3.x.x) and @shopify/react-native-skia (>= 1.2.0)

### Required PRs

1. ✅ The [`MakeImageFromPlatformBuffer(...)` PR in react-native-skia](https://github.com/Shopify/react-native-skia/pull/2344)
2. ⏳ The [`SkData` -> Metal Texture change PR in react-native-skia](https://github.com/Shopify/react-native-skia/pull/2356) (otherwise buffers are in wrong pixelformat and look blue)
3. ⏳ The [YUV CVPixelBuffers PR in react-native-skia](https://github.com/Shopify/react-native-skia/pull/2357) (otherwise we can only use `pixelFormat="rgb"`, no `pixelFormat="yuv"`/default)

### Example

Example shader running directly on the Camera Frame at 60 FPS:

https://github.com/mrousavy/react-native-vision-camera/assets/15199031/e324eed8-eb4b-4af2-94bf-7197d73d14ef

## Changes

- Adds a new API: `useSkiaFrameProcessor` (or `createSkiaFrameProcessor`)
- Marks the Java `FrameProcessorPlugin`'s `callback` function throwable (`throws Throwable`), you can now throw exceptions in Java Frame Processors
- Adds `getPlatformBuffer()` to `Frame`
- Adds a new prop to the `<Camera>`: `preview={boolean}`. This allows you to manually disable the Preview if not needed. (e.g. when using Skia it will use a Skia Canvas instead of the native preview output)
- Fixes an issue where the camera outputs/cameraId are being accessed before they are initialized (`device/no-device` or `session/no-outputs` error) by early aborting duplicate `configure { ... }` calls.
- Minor improvements around the native `Frame` object synchronizations
- Fixes an issue where the Frame Processor pipeline would just stall after a second on Android (forgot to `close()` the parent `ImageProxy`, whoops)
- Fixes an issue where `Frame.orientation` was wrong on iOS (it is always `portrait` since we rotate buffers accordingly in the pipeline)
- Removes the `"native"` pixel format, native does not exist in a Frame Processor - it's always RGB or YUV
- Removes the `supportedPixelFormats` prop on a Format, both iOS and Android always support YUV and RGB, you are free to choose.
- Throws more precise error messages when a YUV pixel format (HDR or compression) cannot be found on iOS, and automatically fallback to the best format to avoid blackscreens
- Adds a `jsi::Scope` to Frame Processor calls to maybe improve GC calls
- Adds lazy proxy dependencies on react-native-worklets-core, react-native-skia and react-native-reanimated

## Tested on

- iPhone 15 Pro (yuv + rgb)
- iPhone 8 (yuv + rgb)
- Huawei P10 (rgb)
- Samsung S22 (rgb)

## Related issues

- Resolves https://github.com/mrousavy/react-native-vision-camera/issues/2734
- Resolves https://github.com/mrousavy/react-native-vision-camera/issues/2347
- Resolves https://github.com/mrousavy/react-native-vision-camera/issues/1755
- Resolves https://github.com/mrousavy/react-native-vision-camera/issues/724
- Resolves https://github.com/mrousavy/react-native-vision-camera/issues/2274
- Resolves https://github.com/mrousavy/react-native-vision-camera/issues/1215

### CODE REVIEW

**Comment by rodgomesc** [comment]
i'm very hyped for this feature!, fantastic work dude 🚀🚀 fun fact: i'm learning a lot with vision camera source code💜💜

**Comment by mrousavy** [comment]
thanks man!

File: package/ios/Frame Processor/FrameHostObject.mm [hunk 1]
```diff
@@ -94,6 +151,30 @@
...
+    auto getPlatformBuffer = JSI_FUNC {
+      // Lock Frame so it cannot be deallocated while we access it
+      std::lock_guard lock(this->_mutex);
+
+      // Box-cast to uintptr (just 64-bit address)
+      CMSampleBufferRef buffer = this->frame.buffer;
+      uintptr_t pointer = reinterpret_cast<uintptr_t>(buffer);
+      jsi::HostFunctionType deleteFunc = [=](jsi::Runtime& runtime, const jsi::Value& thisArg, const jsi::Value* args,
+                                             size_t count) -> jsi::Value {
+        // no-op
```
Comment (wcandillon) [inline package/ios/Frame Processor/FrameHostObject.mm]: no     CFRelease(buffer); here?

File: package/src/dependencies/SkiaProxy.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+type TSkia = typeof Skia
+
+/**
+ * A proxy object that lazy-imports @shopify/react-native-skia as soon as the
+ * caller tries to access a property on {@linkcode SkiaProxy}.
+ *
+ * If @shopify/react-native-skia is not installed, accessing anything on
+ * {@linkcode SkiaProxy} will throw.
+ */
+export const SkiaProxy = createModuleProxy<TSkia>('@shopify/react-native-skia', () => require('@shopify/react-native-skia'))
```
Comment (wcandillon) [inline package/src/dependencies/SkiaProxy.ts:12]: omg this is cool, much better than we came up with for our lose Reanimated coupling.

File: package/ios/Frame Processor/FrameHostObject.mm (hunk 1 above)
Comment (mrousavy) [reply package/ios/Frame Processor/FrameHostObject.mm]: No, I agree that this might be a bit confusing but in my case the Frame is only valid as long as the `useFrameProcessor` didn't return yet. And I `.dispose()` the SkImage before it returns, so there's no way to use it outside of it's lifetime.

Originally I wanted to use HostObjects or NativeState for this so it is properly ref-counted, but that has a few downsides:
1. It requires explicit typings, so react-native-vision-camera needed to depend on react-native-skia in native code, or vice versa so I can use the C++ types.
2. It adds overhead since we'd be using shared_ptrs. (Locks/Mutexes + Ref count increments)
3. It is unnecessary since the Frame cannot live longer than the `useFrameProcessor` anyways (and shouldn't, because that'd stall the Camera pipeline)

File: package/src/dependencies/SkiaProxy.ts (hunk 2 above)
Comment (mrousavy) [reply package/src/dependencies/SkiaProxy.ts:12]: Haha yea thanks I was also pretty excited when I figured out I could just use Proxies for that. I was a bit worried that Hermes force-included those `require`s but it works!

File: package/ios/Frame Processor/FrameHostObject.mm (hunk 1 above)
Comment (wcandillon) [reply package/ios/Frame Processor/FrameHostObject.mm]: Yes that makes sense.

**Comment by mrousavy** [comment]
PR is ready 🎉 
I'll wait until https://github.com/Shopify/react-native-skia/pull/2356 and https://github.com/Shopify/react-native-skia/pull/2357 are merged so we can make use of metal texture conversion and YUV formats, then merge this PR and create a beta release 👍

**Comment by rodgomesc** [comment]
i'm facing a weird bug in the preview for my **mi 9t pro**

-  changing the aspect ratio seems to affect the issue
-  if i comment `frameProcessor={frameProcessor}` everything works fine 
-  if aspectRatio==1 the frame rate seems to be very stable at 30fps

> note: this is a old device, my camera do not support 60fps

<table>
  <tr>
    <td align="center">
     <code>videoAspectRatio: SCREEN_HEIGHT / SCREEN_WIDTH</code>
      <a href="https://github.com/mrousavy/react-native-vision-camera/assets/4893591/35191ad1-e622-4cbd-8a5a-4f03c35965d0">
        
      </a>
    </td>
    <td align="center">
     <code>videoAspectRatio: 1                           </code>
      <a href="https://github.com/mrousavy/react-native-vision-camera/assets/4893591/b9558ff7-6f5c-43e9-9df0-27596a630bd5">
        
      </a>
    </td>
  </tr>
</table>

**Comment by rodgomesc** [comment]
Tested on: `iPhone 14 Pro Max`  
Working like a charm. (Didn't merged the pending YUV stuff from Skia pr for the test.)

<a href="https://github.com/mrousavy/react-native-vision-camera/assets/4893591/8d097a5c-7064-41f3-a50f-b3e4f3e7910c">
  
</a>

**Comment by llli0v0** [comment]
Cool!

**Comment by mrousavy** [comment]
> if aspectRatio==1 the frame rate seems to be very stable at 30fps

Probably just uses a much lower resolution then for thhe Frame. @rodgomesc

**Comment by hashhirr** [comment]
damn! took me three months to make cameras in native modules using kotlin and swift after i found out vision don't have filters...but i guess it will help me in the long run

**Comment by meleffendi** [comment]
I'm trying to overlay a counter on the recorded video. What's the recommended way of doing this at the moment?

**Comment by mrousavy** [comment]
Because of the issue described in https://github.com/Shopify/react-native-skia/pull/2368, I have to make sure to cache my `SkSurface`s on a per-thread basis, as my Frame Processor runs on an iOS DispatchQueue which by design uses multiple C++ Threads.

Unfortunately react-native-skia contexts are tied to C++ threads (`thread_local`), so that's the workaround I am now using: https://github.com/mrousavy/react-native-vision-camera/blob/2b16bf600d52dae7880d0666a3cf191bb26942b3/package/src/skia/useSkiaFrameProcessor.ts#L96-L135

In my short tests, a `DispatchQueue` uses between 8 and 12 Threads, so this creates between 8 and 12 `SkSurface`s:

```
 LOG  Creating new 540x960 Surface...
 LOG  Creating new 540x960 Surface...
 LOG  Creating new 540x960 Surface...
 LOG  Creating new 540x960 Surface...
 LOG  Creating new 540x960 Surface...
 LOG  Creating new 540x960 Surface...
 LOG  Creating new 540x960 Surface...
 LOG  Creating new 540x960 Surface...
```

I agree that the ideal solution is not the one proposed in https://github.com/Shopify/react-native-skia/pull/2368 (the `dispatch_queue_set_specific` approach), but rather an approach where we can create and manage a `SkiaContext` in JS, which I can then pass down to the `SkImage` and `SkSurface` functions (as proposed in https://github.com/Shopify/react-native-skia/pull/2373).
However, this is a **huge** refactor and requires careful construction to make non-breaking, and also I am simply not sure how to get the Context into the DOM/View APIs.

Anyways, my approach works for now, but I'd love to find a better solution for this at some point in the future so I can remove the 7-11 unnecessary `SkSurface`s and get down to just 1 `SkSurface` in memory (cleaner code, and more importantly, much less RAM usage).

**Comment by mrousavy** [comment]
Here's a performance comparison between the two approaches:

|                               | CPU usage | RAM usage |
|-------------------------------|-----------|-----------|
| `thread_local`                | 47%       | 411 MB    |
| `dispatch_queue_set_specific` | 45%       | 334 MB    |

As you can see, the `dispatch_queue_set_specific` approach uses 77 MB less than the approach where I create one `SkSurface` per Thread, and uses 2% CPU less.

So I think we definitely want to avoid making SkContext `thread_local`, and instead re-use it throughout the DispatchQueue.

File: package/src/dependencies/ModuleProxy.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+          target.module = getModule() as TModule
+        } catch (e) {
+          throw new Error(`${name} is not installed!`)
+        }
+      }
+      return target.module[property as keyof typeof holder.module]
+    },
+  })
+  return proxy as unknown as TModule
+}
```
Comment (wcandillon) [inline package/src/dependencies/ModuleProxy.ts:23]: this is cool 😁

**Comment by mrousavy** [comment]
released in latest V4 beta :)

everyone can try `useSkiaFrameProcessor` now in their apps

**Comment by efstathiosntonas** [comment]
@mrousavy I installed beta 15 and it requires Skia to be installed. Can you please consider on making it optional if possible? I guess not everyone is planning on using Skia processors and bloating the app with [6MB](https://shopify.github.io/react-native-skia/docs/getting-started/installation/#bundle-size) is not optimal. 

~edit: by looking at the changes it seems impossible to avoid it~

**Comment by mrousavy** [comment]
Skia is optional: https://github.com/mrousavy/react-native-vision-camera/blob/b751f2da0b1d1543e417a04f0e3b2b077eaa6cc0/package/package.json#L120-L122

Are you seeing an error or anything?

**Comment by mrousavy** [comment]
I have react-native-skia in `devDependencies`: https://github.com/mrousavy/react-native-vision-camera/blob/b751f2da0b1d1543e417a04f0e3b2b077eaa6cc0/package/package.json#L92

Maybe this is why it gets included in your app as well then? I just need the types, that's why I have it in `devDependencies`...

I don't think this ends up in your release app though, as the `devDependencies` are compiled out. Not sure on this.

**Comment by efstathiosntonas** [comment]
after installation I get red screen that skia is missing

**Comment by efstathiosntonas** [comment]
more info:

<details>
  <summary>Click me</summary>
  
Simulator Screenshot - iPhone 13 - 2024-04-17 at 18 56 57

</details>

**Comment by geekabhinavp** [comment]
Hey @mrousavy, can this be merged now, i see all the required PR being merged in Skia.

**Comment by mrousavy** [comment]
@efstathiosntonas fixed in https://github.com/mrousavy/react-native-vision-camera/pull/2750 👍

**Comment by geekabhinavp** [comment]
Hey @mrousavy , I was also getting the skia missing error, i installed it manually in the package. Now that error is gone but I am getting stuck on app loader itself on launch, in the logs i am just getting:
```
iOS Bundling complete 5644ms
[appname] VisionCameraProxy: Creating Worklet Context...
[appname] VisionCameraProxy: Worklet Context Created!
```

**Comment by mrousavy** [comment]
Are there any native logs?

**Comment by geekabhinavp** [comment]
These are the complete logs:
```
[UIKitCore] Class CKBrowserSwitcherViewController overrides the -traitCollection getter,
which is not supported. If you're trying to override traits, you must use the
appropriate API.
iOS ./index.js ░░░░░░░░░░░░░░░░  0.1% ( 2/57)[CoreFoundation] AddInstanceForFactory: No factory registered for id <CFUUID
0x6000006c06a0> F8BB1C28-BAE8-11D6-9C31-00039315CD46
iOS Bundling complete 5644ms
[appname] VisionCameraProxy: Creating Worklet Context...
[appname] VisionCameraProxy: Worklet Context Created!
```
